Twilio-org/phonebank
# Record call attempt

## Manual Testing
1. Do `npm run db:refresh` to empty your database, recreate tables, and seed them with data.
2. Run the following curl command to authorize your user and get the user id and the a token. Note them both so you can use them in curl commands given for this process. Throughout this process assume that the number in the position of `:id` in `/users/:id` for the beginning of any given route is using this id, and [JWT TOKEN HERE] should be replaced with the token returned.
```curl
curl -X POST \
  http://localhost:3000/auth/login \
  -H 'cache-control: no-cache' \
  -H 'content-type: application/x-www-form-urlencoded' \
  -H 'postman-token: 587da4a3-36fb-f0c9-1948-46f8d70853c3' \
  -d 'email=user%40foobar.com&password=password'
```
3. Run the following CURL command with the appropriate user id an token added to get a call and note its id:
```curl
curl -X POST \
  http://localhost:3000/users/1/campaigns/1/calls \
  -H 'authorization: JWT [JWT TOKEN HERE]' \
  -H 'cache-control: no-cache' \
  -H 'postman-token: 23d17009-325e-8b58-ab3b-8b5b9c5dc9d1'
```
4. Put the token user id, call id, and token in the appropriate place in this curl command's route and run it:
```curl
curl -X PUT \
  http://localhost:3000/users/1/campaigns/1/calls/1 \
  -H 'authorization: JWT [JWT TOKEN HERE]' \
  -H 'cache-control: no-cache' \
  -H 'content-type: application/json' \
  -H 'postman-token: 3dccbe1a-b0be-cfd2-739e-ade459cfb1d7' \
  -d '{
    "outcome": "DO_NOT_CALL",
    "notes": null
}'
```
Repeat steps 3 and 4 with different outcomes from `['ANSWERED', 'BAD_NUMBER',  'DO_NOT_CALL', 'LEFT_MSG',  'NO_ANSWER', 'INCOMPLETE']`. Note the following things occur:

- Any attempt recorded with the status `'ANSWERED'` returns status 200 with a json object`{ message: 'call log successfully updated' }`.
- Any attempt recorded with status `'BAD_NUMBER'` or `'DO_NOT_CALL'` returns status 200 with a json object `{ message: 'contact and call log updated successfully.' }`
- Any attempt recorded with status `'LEFT_MSG'`, `'NO_ANSWER'`, `'INCOMPLETE'`  returns stats 200 with a json object `{ message: 'call log successfully updated & call requeued.' }`
- Any attempt recorded with an attempt_num === 3 returns status 200 with a json object`{ message: 'call log successfully updated' }`.
- Attempts recorded with outcomes not listed above return status 400 with a json object `{ message: 'Outcome is not valid' }`
- Attempts recorded on calls that do not have the value `ASSIGNED` on the `status` property returns status 400 with a json object `{ message: 'call has not been assigned' }`.
- Attempts recorded on calls belonging to a campaign for which the user has not joined return status 401 with a json object `{ message: 'User has not joined that campaign' }`.

Note, if you run out of calls and get status 404 with json object `{ message: 'no calls available' }` run `npm run db:refresh` again. This will reset the state of the database of course so any state-mutated dependent conditions will not be true any more.
